source/common/filter/tcp_proxy.cc

```diff
@@ -38,8 +38,15 @@ TcpProxyConfig::Route::Route(
 
 TcpProxyConfig::TcpProxyConfig(const envoy::api::v2::filter::network::TcpProxy& config,
                                Server::Configuration::FactoryContext& context)
-    : stats_(generateStats(config.stat_prefix(), context.scope())),
-      max_connect_attempts_(PROTOBUF_GET_WRAPPED_OR_DEFAULT(config, max_connect_attempts, 1)) {
+    : stats_scope_(context.scope().createScope(fmt::format("tcp.{}.", config.stat_prefix()))),
+      stats_(generateStats(*stats_scope_)),
+      max_connect_attempts_(PROTOBUF_GET_WRAPPED_OR_DEFAULT(config, max_connect_attempts, 1)),
+      upstream_flush_timeout_(std::chrono::seconds(1)), // TODO plumb in real config
```

Comment:
Instead of a new (obscure) config value, should we use https://github.com/envoyproxy/data-plane-api/blob/2f7459baf7c95cf7e634f848948a022c4b16fa79/api/filter/network/tcp_proxy.proto#L36 and time it out if no bytes have been drained in that period?  This would also require a new function on Connection to ask how many bytes it has in the write buffer, but that should be straightforward.

Reply:
+1, it would be great to get that implemented. Not sure if you want to do that first or just use that here and then implement the timer for the entire connection?